librz/analysis/arch/hexagon/hexagon_il.c

```diff
@@ -0,0 +1,783 @@
+			}
+
+			// Low sub-instructions
+			pos->bin.sub[1]->il_insn = hex_il_getter_lt[pos->bin.sub[1]->identifier];
+			cur_il_insn = &pos->bin.sub[1]->il_insn;
+			// low sub operation 0
+			cur_il_insn->op0.hi = pos->bin.sub[1];
+			if (cur_il_insn->op0.attr == HEX_IL_INSN_ATTR_INVALID) {
+				goto not_impl;
+			}
+			rz_vector_push(p->il_ops, &cur_il_insn->op0);
+
+			// low sub operation 1
+			if (cur_il_insn->op1.attr != HEX_IL_INSN_ATTR_INVALID) {
+				pos->bin.sub[1]->il_insn.op1.hi = pos->bin.sub[1];
+				rz_vector_push(p->il_ops, &cur_il_insn->op1);
+			}
+		} else {
+			pos->bin.insn->il_insn = hex_il_getter_lt[pos->bin.insn->identifier];
+			cur_il_insn = &pos->bin.insn->il_insn;
+			// Insn operation 0
+			cur_il_insn->op0.hi = pos->bin.insn;
+			if (cur_il_insn->op0.attr == HEX_IL_INSN_ATTR_INVALID) {
+				goto not_impl;
+			}
+			rz_vector_push(p->il_ops, &cur_il_insn->op0);
+			// Insn operation 1
+			if (cur_il_insn->op1.attr != HEX_IL_INSN_ATTR_INVALID) {
+				cur_il_insn->op1.hi = pos->bin.insn;
+				rz_vector_push(p->il_ops, &cur_il_insn->op1);
+			}
+		}
+	}
+	return true;
+not_impl:
+	RZ_LOG_INFO("Hexagon instruction %" PFMT32d " not implemented.\n", pos->bin.insn->identifier);
+	return false;
+}
+
+static void check_for_jumps(const HexPkt *p, RZ_OUT bool *jump_flag) {
+	rz_return_if_fail(p && jump_flag);
+	HexILOp *op;
+	rz_vector_foreach(p->il_ops, op) {
+		if (op->attr & HEX_IL_INSN_ATTR_BRANCH) {
+			*jump_flag = true;
+		}
+	}
+}
+
+/**
+ * \brief Checks if the packet at \p addr has all conditions fulfilled
+ * to be executed.
+ *
+ * \param pkt The packet to check.
+ * \param addr Address of the requested IL operation.
+ *
+ * \return true If the packet can be set up for emulation.
+ * \return false Otherwise.
+ */
+static inline bool pkt_at_addr_is_emu_ready(const HexPkt *pkt, const ut32 addr) {
+	if (rz_list_length(pkt->bin) == 1) {
+		const HexInsnContainer *hic = rz_list_get_top(pkt->bin);
+		if (hic->identifier == HEX_INS_INVALID_DECODE) {
+			return false;
+		}
+	}
+	return addr == pkt->pkt_addr && pkt->is_valid && pkt->last_instr_present;
+}
+
+/**
+ * \brief Returns the IL operation of the instruction at \p addr. This will always be EMPTY().
+ * Except for last instructions in a packet. Those will always return the complete IL operation
+ * of the packet or NULL if one instruction was not implemented or an error occurred.
+ *
+ * \param addr Address of the requested IL operation.
+ * \param get_pkt_op If true, it returns the IL operation of the whole packet at \p addr.
+ * It will return EMPTY() if there is no packet which starts at \p addr.
+ * If false, the behavior is as documented above.
+ * \return RzILOpEffect* Sequence of operations to emulate the packet.
+ */
+RZ_IPI RzILOpEffect *hex_get_il_op(const ut32 addr, const bool get_pkt_op) {
+	static bool might_has_jumped = false;
+	HexState *state = hexagon_get_state();
+	if (!state) {
+		RZ_LOG_WARN("Could not get hexagon state!\n");
```

Comment:
Maybe "hexagon instruction decoder state" or something like that? I think just "hexagon state" is a bit misleading.